File: src/lib/providers/router.ts
import { LLMProvider, ModelConfig } from './interface';
import { OpenRouterProvider } from './openrouter';
import { GroqProvider } from './groq';
import { InceptionProvider } from './inception';
import {
  GenerateTextOptions, GenerateTextResult,
  StructuredExtractOptions, StructuredExtractResult,
  RerankOptions, RerankResult,
} from '@/types';

/**
 * ModelRouter manages provider selection with automatic fallback.
 * Routes requests to the best available provider for each pipeline stage.
 */
export class ModelRouter {
  private providers: LLMProvider[] = [];
  private config: ModelConfig;

  constructor(config?: Partial<ModelConfig>) {
    const openrouter = new OpenRouterProvider();
    const groq = new GroqProvider();
    const inception = new InceptionProvider();
    this.providers = [openrouter, groq, inception].filter(p => p.isAvailable());

    this.config = {
      queryClassification: 'openrouter:google/gemini-2.0-flash-lite-preview-02-05:free',
      evidenceSummarization: 'groq:llama-3.3-70b-versatile',
      answerSynthesis: 'openrouter:meta-llama/llama-3.1-8b-instruct:free',
      embedding: 'openrouter:text-embedding-3-small',
    };

    if (process.env.QUERY_CLASSIFICATION_MODEL) this.config.queryClassification = process.env.QUERY_CLASSIFICATION_MODEL;
    if (process.env.EVIDENCE_SUMMARIZATION_MODEL) this.config.evidenceSummarization = process.env.EVIDENCE_SUMMARIZATION_MODEL;
    if (process.env.ANSWER_SYNTHESIS_MODEL) this.config.answerSynthesis = process.env.ANSWER_SYNTHESIS_MODEL;
    if (process.env.EMBEDDING_MODEL) this.config.embedding = process.env.EMBEDDING_MODEL;
  }

  private parseModelSpec(spec: string): { providerName: string; modelName: string } {
    const parts = spec.split(':');
    if (parts.length >= 2) {
      const providerName = parts[0].toLowerCase();
      const modelName = parts.slice(1).join(':');
      return { providerName, modelName };
    }
    return { providerName: 'openrouter', modelName: spec };
  }

  private getProvider(name: string): LLMProvider | undefined {
    return this.providers.find(p => p.name === name);
  }

  private getFirstAvailable(): LLMProvider | undefined {
    return this.providers[0];
  }


  // Provider-specific fallback models
  private fallbackModels: Record<string, string> = {
    openrouter: 'google/gemini-2.0-flash-001:free',
    groq: 'llama-3.3-70b-versatile',
    inception: 'mercury',
  };

  async generateText(options: GenerateTextOptions, stage?: string): Promise<GenerateTextResult> {
    let modelSpec = options.model;
    if (stage && !options.model) {
      if (stage === 'queryClassification') modelSpec = this.config.queryClassification;
      else if (stage === 'evidenceSummarization') modelSpec = this.config.evidenceSummarization;
      else modelSpec = this.config.answerSynthesis;
    }
    const { providerName, modelName } = this.parseModelSpec(
      modelSpec || 'openrouter:google/gemini-2.0-flash-001:free'
    );
    const primaryProvider = this.getProvider(providerName) || this.getFirstAvailable();
    if (!primaryProvider) throw new Error('No LLM providers available');

    const errors: string[] = [];
    for (const provider of [primaryProvider, ...this.providers.filter(p => p.name !== primaryProvider.name)]) {
      // Use provider-specific model name during fallback
      const fallbackModel = provider.name !== primaryProvider.name
        ? (this.fallbackModels[provider.name] || modelName)
        : modelName;
      try {
        return await provider.generateText({ ...options, model: fallbackModel });
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        errors.push(`${provider.name}: ${msg}`);
        continue;
      }
    }
    throw new Error(`All providers failed: ${errors.join('; ')}`);
  }

  async extractStructured(options: StructuredExtractOptions): Promise<StructuredExtractResult> {
    const { providerName, modelName } = this.parseModelSpec(
      options.model || 'openrouter:google/gemini-2.0-flash-001:free'
    );
    const primaryProvider = this.getProvider(providerName) || this.getFirstAvailable();
    if (!primaryProvider) throw new Error('No providers available');

    const errors: string[] = [];
    for (const provider of [primaryProvider, ...this.providers.filter(p => p.name !== primaryProvider.name)]) {
      const fallbackModel = provider.name !== primaryProvider.name
        ? (this.fallbackModels[provider.name] || modelName)
        : modelName;
      try {
        return await provider.extractStructured({ ...options, model: fallbackModel });
      } catch (err: unknown) {
        errors.push(`${provider.name}: ${err instanceof Error ? err.message : String(err)}`);
        continue;
      }
    }
    throw new Error(`All providers failed: ${errors.join('; ')}`);
  }

  async rerankCandidates(options: RerankOptions): Promise<RerankResult> {
    const { providerName, modelName } = this.parseModelSpec(
      options.model || 'openrouter:google/gemini-2.0-flash-lite-preview-02-05:free'
    );
    const primaryProvider = this.getProvider(providerName) || this.getFirstAvailable();
    if (!primaryProvider) throw new Error('No providers available');

    const errors: string[] = [];
    for (const provider of [primaryProvider, ...this.providers.filter(p => p.name !== primaryProvider.name)]) {
      try {
        return await provider.rerankCandidates({ ...options, model: modelName });
      } catch (err: unknown) {
        errors.push(`${provider.name}: ${err instanceof Error ? err.message : String(err)}`);
        continue;
      }
    }
    throw new Error(`All providers failed: ${errors.join('; ')}`);
  }

  getModelForStage(stage: 'queryClassification' | 'evidenceSummarization' | 'answerSynthesis'): string {
    return this.config[stage];
  }

  hasAvailableProviders(): boolean {
    return this.providers.length > 0;
  }
}

let routerInstance: ModelRouter | null = null;

export function getRouter(): ModelRouter {
  if (!routerInstance) routerInstance = new ModelRouter();
  return routerInstance;
}

File: backend/app/providers/router.py
"""Provider router — routes requests to the cheapest/fastest available provider with fallback."""

import logging
from typing import Optional

from app.providers.interface import (
    LLMProvider, GenerateTextOptions, GenerateTextResult,
    StructuredExtractOptions, StructuredExtractResult,
)
from app.providers.groq_adapter import GroqProvider
from app.providers.openrouter_adapter import OpenRouterProvider
from app.providers.inception_adapter import InceptionProvider
from app.config import get_settings

logger = logging.getLogger(__name__)


def _parse_model_ref(ref: str) -> tuple[str, str]:
    """Parse 'provider:model_name' into (provider_name, model_name)."""
    if ":" in ref:
        parts = ref.split(":", 1)
        return parts[0], parts[1]
    return "", ref


class ProviderRouter:
    """Routes LLM requests to the best available provider with automatic fallback."""

    def __init__(self):
        self._providers: dict[str, LLMProvider] = {
            "groq": GroqProvider(),
            "openrouter": OpenRouterProvider(),
            "inception": InceptionProvider(),
        }
        self._settings = get_settings()

    def _get_provider(self, name: str) -> Optional[LLMProvider]:
        provider = self._providers.get(name)
        if provider and provider.is_available():
            return provider
        return None

    def _get_fallback_chain(self, preferred_provider: str) -> list[LLMProvider]:
        """Build fallback chain starting with preferred provider."""
        chain = []
        p = self._get_provider(preferred_provider)
        if p:
            chain.append(p)
        for name, provider in self._providers.items():
            if name != preferred_provider and provider.is_available():
                chain.append(provider)
        return chain

    async def generate_text(
        self,
        options: GenerateTextOptions,
        model_override: Optional[str] = None,
    ) -> GenerateTextResult:
        """Generate text with automatic fallback across providers."""
        if model_override:
            provider_name, model_name = _parse_model_ref(model_override)
        elif options.model:
            provider_name, model_name = _parse_model_ref(options.model)
        else:
            provider_name = ""
            model_name = ""

        chain = self._get_fallback_chain(provider_name)

        if not chain:
            raise RuntimeError("No LLM providers available. Check API keys in .env")

        last_error = None
        for provider in chain:
            try:
                opts = GenerateTextOptions(
                    prompt=options.prompt,
                    model=model_name or None,
                    temperature=options.temperature,
                    max_tokens=options.max_tokens,
                    system=options.system,
                )
                result = await provider.generate_text(opts)
                return result
            except Exception as e:
                last_error = e
                logger.warning("provider=%s failed: %s, trying next", provider.name, e)
                continue

        raise RuntimeError(f"All providers failed. Last error: {last_error}")

    async def extract_structured(
        self,
        options: StructuredExtractOptions,
        model_override: Optional[str] = None,
    ) -> StructuredExtractResult:
        """Extract structured data with automatic fallback."""
        if model_override:
            provider_name, model_name = _parse_model_ref(model_override)
        else:
            provider_name, model_name = _parse_model_ref(options.model)

        chain = self._get_fallback_chain(provider_name)

        if not chain:
            raise RuntimeError("No LLM providers available")

        last_error = None
        for provider in chain:
            try:
                opts = StructuredExtractOptions(
                    text=options.text,
                    schema=options.schema,
                    instructions=options.instructions,
                    model=model_name,
                )
                result = await provider.extract_structured(opts)
                return result
            except Exception as e:
                last_error = e
                logger.warning("provider=%s extract failed: %s", provider.name, e)
                continue

        raise RuntimeError(f"All providers failed for extraction. Last error: {last_error}")

    def available_providers(self) -> list[str]:
        """Return list of available provider names."""
        return [name for name, p in self._providers.items() if p.is_available()]


_router: Optional[ProviderRouter] = None


def get_provider_router() -> ProviderRouter:
    global _router
    if _router is None:
        _router = ProviderRouter()
    return _router
